dask/dataframe/core.py

```diff
@@ -1669,7 +1669,20 @@ def isna(self):
 
     @derived_from(pd.DataFrame)
     def isin(self, values):
-        return elemwise(M.isin, self, list(values))
+        if is_dataframe_like(self._meta):
+            # DataFrame.isin does weird alignment stuff
+            bad_types = (_Frame, pd.Series, pd.DataFrame)
+        else:
+            bad_types = (_Frame,)
+        if isinstance(values, bad_types):
+            raise NotImplementedError(
+                "Passing a %r to `isin`" % typename(type(values))
+            )
+        meta = self._meta_nonempty.isin(values)
+        # We wrap values in a delayed for two reasons:
+        # - avoid serializing data in every task
+        # - avoid cost of traversal of large list in optimizations
+        return self.map_partitions(M.isin, delayed(values), meta=meta)
```

Comment:
I think that we call `delayed` already *if* the object is large.  This happens in `map_partitions` if you look at the normalized_arg call

https://github.com/dask/dask/blob/21f8de898fca9edddb81de388012652ab89da80c/dask/array/core.py#L288-L305

If this isn't happening in some cases then it might make sense to review what's there and change things at that point instead.

(though, to be clear, I think that this change is fine)

Reply:
We were using `elemwise` before, which doesn't have that code path. In this case I think the threshold for nbytes (currently 1e6) in large lists (a common argument type) is too low - the cost of traversing the list in the graph for both optimization and execution becomes expensive with many partitions.

Reply:
Should we improve `normalize_arg` to put any list of length greater than 10 into a task with delayed?

```python
if isinstance(x, list) and len(x) > 10:
    return delayed(x)
```

Mostly I'm just looking for ways to generalize the improvement you've found here.  I suspect that it burns us in other situations as well.

Reply:
Maybe? I'd prefer to experiment with that later - there's a few things I'd like to cleanup in dask-dataframe like this. I'll make an issue.

Reply:
#4734